articles/ros_node_idl.md

```diff
@@ -0,0 +1,249 @@
+
+Considering the high level abstraction of the IDL, the `interface` tag is not meant for a specific package but rather declares intrinsic properties for anyone to process it.
+However, the keyword is likely solidly descriptive enough to either be accepted as falling under [REP 149][rep149_export] or else to motivate an amendment of the REP.
+
+### IDL Schema
+
+#### `interface`
+
+Root tag of the IDL file, it is made up of a collection of node interfaces.
+There must be only one tag per IDL file.
+
+Attributes:
+- **version**: version of schema in use, allowing for future revisions.
+
+#### `node`
+
+Encapsulate a sequence of ROS entity interfaces.
+It is specific to a node as determined by its associated attributes.
+
+Attributes:
+- **name**: The base name of the node.
+
+#### `action`
+
+Define the interface for a given action.
+
+Attributes:
+- **name**: The name of the action.
+- **type**: The type of the action.
+Valid values are any ROS action types.
+- **server**: Whether or not the node provides an action server for the action.
+Valid values are "true" or "false". Defaults to "false".
+- **client**: Whether or not the node provides a client for the action.
+Valid values are "true" or "false". Defaults to "false".
+
+Sub-tag:
+- **qos**: The Quality of Service setting for the action - see the [qos tag](####qos).
+Defaults to [rcl_action_qos_profile_status_default][rcl_action_qos_profile_status_default_link]
+
+#### `parameter`
+
+Define the interface for a given parameter.
+
+Attributes:
+- **name**: The name of the parameter.
+- **type**: The type of the parameter.
+Valid values are any ROS parameter types.
+
+#### `service`
+
+Define the interface for a given service.
+
+Attributes:
+- **name**: The name of the service.
+- **type**: The type of the service.
+Valid values are any ROS service types.
+- **server**: Whether or not the node provides a server for the service.
+Valid values are "true" or "false". Defaults to "false".
+- **client**: Whether or not the node provides a client for the service.
+Valid values are "true" or "false". Defaults to "false".
+
+Sub-tag:
+- **qos**: The Quality of Service setting for the service - see the [qos tag](####qos).
+Defaults to [rmw_qos_profile_services_default][rmw_qos_profile_services_default_link]
+
+#### `topic`
+
+Define the interface for a given topic.
+
+Attributes:
+- **name**: The name of the topic.
+- **type**: The type of the message.
+Valid values are any ROS message types.
+- **publisher**: Whether or not the node publishes on the topic.
+Valid values are "true" or "false". Defaults to "false".
+- **subscription**: Whether or not the node subscribes to the topic.
+Valid values are "true" or "false". Defaults to "false".
+
+Sub-tag:
+- **qos**: The Quality of Service setting for the topic - see the [qos tag](####qos).
+Defaults to [rmw_qos_profile_default][rmw_qos_profile_default_link]
+
+#### `qos`
```

Comment:
Following the discussion [here](https://github.com/ros2/design/pull/266#discussion_r350380468) we opted for a nested object approach to define `qos` (see e.g. [here](https://github.com/ros2/design/pull/266/files#diff-ddd4bb0e2616a5d1c2e7667c294d309eR11)) rather than an attribute as you seem to be suggesting. I agree that it would be handy to be able to re-use those pre-defined profiles somehow. Maybe there could be another `qos` attribute such as follow,
```xml
<qos profile="sensor_data">
```
but then we would have to handle cases such as defining a `profile` and some other options, e.g.,
```xml
<qos profile="sensor_data" history=5>
```
What would we do in such case? Override the `history` param of the `sensor_data` `profile`, raise an exception, something else?